### PR TITLE
build: added separate option for optimization level

### PR DESCRIPTION
**Pull request by hoopoepg**
- build script will not touch optimization level
  by default
- to specify optimization level --enable-fast=level
  may be used

this is fix for https://github.com/openucx/ucx/issues/1696

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4397/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
@shamisp what do you think?

File: config/m4/compiler.m4 [hunk 1]
```diff
@@ -17,8 +17,19 @@ AC_ARG_ENABLE(debug,
...
+#
+# Optimization level
+#
+AC_ARG_ENABLE(fast,
+        AC_HELP_STRING([--enable-fast], [Set optimization level [0-3]]),
+        [],
+        [enable_fast=""])
+AS_IF([test -z "$enable_fast"], [],
+      [test "$enable_fast" == "yes"], [BASE_CFLAGS="-O3 $BASE_CFLAGS"],
+      [BASE_CFLAGS="-O$enable_fast $BASE_CFLAGS"])
```
Comment (miked-mellanox) [inline config/m4/compiler.m4]: now the default behave has changed:

before this PR - O3 was enable by default
after this PR - no optimization

maybe we should have enable_fast=yes by default to match default behavior?

Comment (hoopoepg) [reply config/m4/compiler.m4]: ok, updated...

File: config/m4/compiler.m4 [hunk 2]
```diff
@@ -17,8 +17,20 @@ AC_ARG_ENABLE(debug,
         [],
         [enable_debug=no])
 AS_IF([test "x$enable_debug" == xyes],
-        [BASE_CFLAGS="-O0 -D_DEBUG $BASE_CFLAGS"],
-        [BASE_CFLAGS="-O3 $BASE_CFLAGS"])
+        [BASE_CFLAGS="-D_DEBUG $BASE_CFLAGS"],
+        [])
```
Comment (miked-mellanox) [inline config/m4/compiler.m4]: enable_debug should be mutual exclusive with enable_fast and probably set enable_fast=0
or add check in enable_fast that enable_debug is not set

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4407/ for details (Mellanox internal link).

File: config/m4/compiler.m4 (hunk 2 above)
Comment (hoopoepg) [reply config/m4/compiler.m4]: ok, done

**Comment by shamisp** [comment]
I think by default we should not do any optimization. I think original
default behavior was bad.

**Comment by miked-mellanox** [comment]
i think default should be optimized mode, as we aim for best perf OOB, unless user selects otherwise (debug or level=x)

WDYT

**Comment by shamisp** [comment]
It is a bad practice to push "o" optimization in a default build

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4418/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
bot:bgate:retest

**Comment by hoopoepg** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4423/ for details (Mellanox internal link).

File: config/m4/compiler.m4 [hunk 3]
```diff
@@ -17,8 +17,27 @@ AC_ARG_ENABLE(debug,
...
+
+#
+# Optimization level
+#
+AC_ARG_ENABLE(fast,
+        AC_HELP_STRING([--enable-fast], [Set optimization level [0-3]]),
+        [],
+        [enable_fast="none"])
+AS_IF([test -z "$enable_fast"], [],
+      [test "$enable_fast" == "yes"], [BASE_CFLAGS="-O3 $BASE_CFLAGS"],
```
Comment (yosefe) [inline config/m4/compiler.m4]: so by default UCX will be compiled without optimizations?
what is the reason for this change?

Comment (hoopoepg) [reply config/m4/compiler.m4]: No, default behavior will not be changed

Comment (yosefe) [reply config/m4/compiler.m4]: maybe i'm missing something, but by default enable_fast will be empty, which means `[test -z "$enable_fast"]` is true, so no optimization flags will be added...

Comment (hoopoepg) [reply config/m4/compiler.m4]:
```
AC_ARG_ENABLE(fast,
        AC_HELP_STRING([--enable-fast], [Set optimization level [0-3]]),
        [],                   >>>>>>>>>>>>>> here is action when enable_fast is set
        [enable_fast="none"]) >>>>>>>>>>>>>> here is action when enable_fast is not set
```
in case if user doesn't set enable_fast then optimization level depends from ```enable_debug``` variable (below, line 36)

File: config/m4/compiler.m4 [hunk 4]
```diff
@@ -17,8 +17,27 @@ AC_ARG_ENABLE(debug,
...
+      [test "$enable_fast" == "none"],
+          [AS_IF([test "x$enable_debug" == xyes],
```
Comment (hoopoepg) [inline config/m4/compiler.m4]: here is decision of what optimization level should be depends from debug level

File: config/m4/compiler.m4 (hunk 3 above)
Comment (yosefe) [reply config/m4/compiler.m4]: i see. so what is the purpose of `test -z "$enable_fast`? seems like this variable can never be empty - if it's given, it's set to yes/no, and if it's not given it's set to none?
maybe replace "enable_fast=none" by "enable_fast=yes"?

Comment (hoopoepg) [reply config/m4/compiler.m4]: technically it is possible, for example if we decide to disable optimization settings based on another flags (for example for any "exotic" compiler), in this case we have to set to enable_fast empty value to suppress -Ox parameters, in other cases this is unnecessary case.

so, remove it? (at least it will simplify ligic)

Comment (yosefe) [reply config/m4/compiler.m4]: yes, i think it's better to simplify logic and for now work with enable_fast which can be only "yes"
 (default) or "no" (if ./configure --disable-fast was done")
BTW is "disable-fast" a standard terminology? maybe rename it to "disable-compiler-opt" or something else?

Comment (hoopoepg) [reply config/m4/compiler.m4]: done: removed empty switch, argument renamed to disable-compiler-opt

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4497/ for details (Mellanox internal link).

File: config/m4/compiler.m4 [hunk 5]
```diff
@@ -17,8 +17,26 @@ AC_ARG_ENABLE(debug,
...
+
+#
+# Optimization level
+#
+AC_ARG_ENABLE(compiler-opt,
+        AC_HELP_STRING([--enable-compiler-opt], [Set optimization level [0-3]]),
+        [],
+        [enable_compiler_opt="none"])
+AS_IF([test "$enable_compiler_opt" == "yes"], [BASE_CFLAGS="-O3 $BASE_CFLAGS"],
+      [test "$enable_compiler_opt" == "none"],
```
Comment (yosefe) [inline config/m4/compiler.m4]: why do we need special case for 'none'? seems like we can say that the default value of "enable_compiler_opt" can be "yes" rather than none. Since in both cases, "enable_debug=yes" would override "enable_compiler_opt" to "-O0"

Comment (hoopoepg) [reply config/m4/compiler.m4]: 'none' is special case when user didn't set any value, in this case it sets optimization level based on debug configuration. in case if user set level (```--enable-compiler-opt``` or ```--enable-compiler-opt=value```) system forces specified level. The "feature" here is set O3 for release and O0 for debug by default, but force custom level in case if user wants it

Comment (yosefe) [reply config/m4/compiler.m4]: so why is the 'if' in line 22 needed?

Comment (hoopoepg) [reply config/m4/compiler.m4]: artefact (to prevent out-of-ordering variable settings)
removed

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/4501/ for details (Mellanox internal link).

**Comment by hoopoepg** [comment]
bot:bgate:retest

**Comment by shamisp** [comment]
If no special flag is passed, we should not force -03.

**Comment by hoopoepg** [comment]
Why not? If user decide to disable optimization, or set own level system will respect it

**Comment by yosefe** [comment]
@shamisp the current behavior is to compile with -O3 by default, unless external CFLAGS are given - for best out-of-box performance. This PR maintains this behavior.

**Comment by yosefe** [comment]
bot:ornl:retest

**Comment by shamisp** [comment]
I think the default behavior was wrong - overwriting optimization flags out of the box is not a good idea. We have config script in contrib, this should do the job.

**Comment by yosefe** [comment]
@shamisp the flags are not overridden by default. if no flags are given, the default is -O3. And this is good because we don't want that users would have to provide CFLAGS if they want decent performance.

**Comment by shamisp** [comment]
We have been there. You do overwrite this you append to existing flags. I'm not sure if there is guaranteed order in which it is processed.

**Comment by yosefe** [comment]
@shamisp we set BASE_CFLAGS, not CFLAGS. same way that is done in #1555, which you approved. what has changed?

**Comment by yosefe** [comment]
@shamisp ?

**Comment by shamisp** [comment]
I still think this is a silly and unnecessary option.  Why do we need special configure option to do ./configure ... CFLAGS=-03 ? If you want to automate it, it should be handled through configure scripts in contrib.

**Comment by shamisp** [comment]
Discussed with @yosefe offline. I'm okay with the patch, even so I don't like it :-)
